### ISSUE TITLE
[Feature Request] Spreadsheets embedded in notes like notion

### ISSUE THREAD

**Issue by surajsharma** [zadam/trilium#1841]
For bug reports, **PLEASE mention version of Trilium you're using** and also include **log files** from following location:

* `/home/[user]/.local/share/trilium-data/log` for Linux
* `C:\Users\[user]\AppData\Roaming\trilium-data\log` for Windows Vista and up
* `/Users/[user]/Library/Application Support/trilium-data/log` for Mac OS

**Comment by zadam** [comment]
Hi, I have thought about that since yeah, it would be pretty cool. Developing spreadsheet is probably way too much to do in this project so we'd have to use some existing library. There's a bunch of them with a range of features and licenses. Some of them are pretty small and light (e.g. [jSpreadsheet](https://github.com/jspreadsheet/ce)), some of them very large and complex (AG Grid).

It would be good to start with thinking about which features are we actually targetting.

**Comment by surajsharma** [comment]
oh jSpreadsheet used to be called jExcel, they have since changed their name, I have used this in another project, yes it is pretty small and efficient at what it does.

**Comment by aihuasxy** [comment]
Luckysheet is another alterative js sheet, looks cool!
https://github.com/mengshukeji/Luckysheet

**Comment by surajsharma** [comment]
definitely a great find, how could one go about adding it in? if you could give a basic outline, I will try and make a PR

**Comment by chee** [comment]
Will a spreadsheet be a new type of note? And will the data from it be usable in a script note?

**Comment by surajsharma** [comment]
Yes, and Yes ... i think?

Cordially,
Suraj @ [Evenzero.in](http://evenzero.in/)
[LinkedIn](https://www.linkedin.com/in/surajsharma21), [Github](https://github.com/surajsharma/), [StackOverflow](https://stackoverflow.com/users/11857517/suraj-sharma), [CV](https://bit.ly/3rFa9UB), [Projects](http://evenzero.in/projects/), [Experiments](http://evenzero.in/experiments)
--

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by zadam** [comment]
> Will a spreadsheet be a new type of note

Probably yes even though it does also have disadvantages. On one hand I like how e.g. [CKeditor](https://ckeditor.com/docs/ckeditor4/latest/features/spreadsheets.html) does it in their paid plugin that it integrates into the larger document. On the other hand especially for e.g. script processing it's better to have tables separate. It might be possible to still include the table from a separate note into the text document via "include note" but then it will be probably read only in the text document and editable only from the spreadsheet note.

> And will the data from it be usable in a script note?

Ideally the spreadsheet data would be JSON and thus easy to process in the script.

> definitely a great find, how could one go about adding it in? if you could give a basic outline, I will try and make a PR

Experimentation is great, but otherwise I'm not likely to accept such a PR so quickly. For me adding a spreadsheet plugin is a big commitment regarding the features, data format, long term viability. So I'd like to do some serious research into available libraries first and figure out what would be the best.

**Comment by martinkoutecky** [comment]
One thing I love about notion's spreadsheets is that they are light-weight databases, so there's a column type which links to some other table, and one selects the values in this column to be rows of the linked table. (Similar as Airtable or Seatable.) This is incredibly helpful and once you see it, quite natural (and addictive). Not sure any of the mentioned plugins have this kind of functionality, and how it would play out in Trilium, but it would be super cool to have something like that.

**Comment by surajsharma** [comment]
> and how it would play out in Trilium

oh thats nice! i didn't know Notion had such a feature. Trilium could certainly do with this! However, it might not be for the same reasons or same purposes even. The idea here is not to make Trilium more like Notion but to augment the feature list, in a way that is consistent with the software's essence. Honestly, when I opened this issue, I was merely looking for a simple spreadsheet note type without the bells and whistles.

**Comment by chee** [comment]
I think that relates to this issue: https://github.com/zadam/trilium/issues/822

**Comment by zadam** [comment]
> One thing I love about notion's spreadsheets is that they are light-weight databases, so there's a column type which links to some other table, and one selects the values in this column to be rows of the linked table. (Similar as Airtable or Seatable.)

If I understand this correctly then this is supported by some of the more advanced plugins within one spreadsheet between different lists (so just like in Excel). 

My goals here are rather modest - the chosen plugin should work in Trilium mostly as designed and its integration into Trilium should be mainly a question of data storage (versioning, sync etc.) I don't plan advanced integration between spreadsheet and rest of Trilium, even basic things like being able to link a note or image from a spreadsheet will be probably challenging. This also depends on the plugin architecture - i.e. is it itself pluggable or do I need to hard-fork it to change things?

**Comment by surajsharma** [comment]
> I don't plan advanced integration between spreadsheet and rest of Trilium, even basic things like being able to link a note or image from a spreadsheet will be probably challenging. 

indeed, but i think spreadsheet-to-spreadsheet connectivity should still be relatively less complex to implement. i really don't like notion for some odd reason even though all these entrepreneur types swear by it these days...

**Comment by Rsose** [comment]
Wow great idea!
If we can directly implement some simple math formulas in Trilium Notes inside that would be very nice.
Optimize the Table function by adding a spreadsheet to it, so that the data table can be fully implemented in Trilium Notes.
Just think of using Trilium tables to generate all kinds of calculation data tables(or another way to realize it),it will be better to manage my spreadsheets, cos I use excel to generate them, the main problem is that the files is more difficult to manage up.
Sometimes I generate a lot of excel files, after a while it is easy to forget where they put!
I recently wanted to submit a similar ISSUE and found that it was mentioned in this question.
If there is no spreadsheet in Trilium, I need to generate it from Excel and import it to trilium.That's a lot of steps, not convenient.
Of course, the best choice is to put them in trilium 👍

**Comment by lucamarletta** [comment]
I'd love this feature too. 
I'd like as well the possibility to save it as all formats as for other notes,: md, pdf ecc..
I could take notes of some simple bills, just like that. If one needs a complex table with Math formulas can just open LibreOffice.

**Comment by good-praxis** [comment]
For the time being, is there a way to programmatically generate a dynamic table in CKEditor? Say, I want a table that aggregates all child notes with their promoted attributes visible to compare? although sorting on click and the like is out of the question

**Comment by zadam** [comment]
@good-praxis [Scripting](github.com/zadam/trilium/wiki/Scripts) would be a way to achieve this.

**Comment by finzzz** [comment]
Rather than reinventing new stuff, is it possible to integrate it with something like nocodb or baserow?

**Comment by toshi2d** [comment]
Any updates?

**Comment by TheOneValen** [comment]
> Rather than reinventing new stuff, is it possible to integrate it with something like nocodb or baserow?

Intergation with nocodb. e.g. via the api would be awesome.

**Comment by jwest75674** [comment]
My use case would benefit from the ability to add a spreadsheet into an existing text note, like how we can add a table of varying row counts and column counts. 

Specifically, when taking notes, I often need to do some simple math to keep everything clear later of how I can do specific numbers or with a calculator open independently. It would be great to get the math done inside a table!

In its simplest form, adding the ability to do basic arithmetic to the existing table functionality would be enough for my needs, and could avoid some of the complexity some of the previously proposed paths would likely entail. Basic stuff, multiplication, division, addition, subtraction, etc.

**Comment by zadam** [comment]
Recently also disccused in https://github.com/zadam/trilium/issues/410

**Comment by kalmigs** [comment]
Hi, is there an ongoing discussion to add spreadsheets? if so can anyone link?

**Comment by eliandoran** [comment]
@kalmigs 

Nope. We haven't found a viable technology to use for spreadsheets. The table collection is the only thing we have so far.

**Comment by kalmigs** [comment]
@eliandoran thank you for replying. Really appreciate your work on the project!

Do you mean:
- currently, no spreadsheet JS projects meet the requirements?
- Or it's harder to integrate those into trilium because of the way it is architected?

**Comment by eliandoran** [comment]
@kalmigs , I would say the architecture of Trilium is now at its highest point since we have ported to React. We didn't find any suitable candidate for integrating, however. 

The only possible candidate is https://github.com/dream-num/univer/ but I'm not sure if it's a good fit. We have to consider how maintained the application is and who are the developers, so that we don't accidentally vendor lock ourselves in.

There's also the problem that not everyone needs a spreadsheet so perhaps a plugin system would be better to have first.

**Comment by kalmigs** [comment]
Yeah I agree that a plugin system will be the best approach.. thanks!

**Comment by rom1dep** [comment]
It's probably good to spend a little bit of time defining "spreadsheet",
AFAICT, it's pretty broad and under-defined:

- structured content:
  - the ability to render data as a table → we've got this base covered with collections
  - the ability to conveniently "bulk"-edit the data (range edit, multi-cell copy-paste, fill-down, etc) → there's probably much more we can do with the table component that we use right now
  - the ability to perform calculations/formulas → I hope that one day we might have a concept of "calculated attributes" whose values can be inferred from others (something like `#color=${if(self.priority === "urgent") "red"} #overDue=${if(self.dueDate <= $TODAY) true else false}`
- unstructured content:
  - actual spreadsheet-like "sea of cells" and rich content like formatting, images, … → at that rate, it makes more practical sense to embed an actual spreadsheet solution like collabora-online and treat the output as a blob stored as note attachment (with very little trilium knows or cares about what's inside)

**Comment by thfrei** [comment]
Defining spreadsheet is a good idea.

Because, rather than a spreadsheet, I keep thinking about the notion "database" feature.
I use Notion a lot at work. The thing that keeps people hooked, and that only notion and not it competitors offer, is the database feature and being able to link databases with each other. This is quite powerful and allows for a lot of business use cases. 
So i'd say one note type being kind of a database (table/spreadsheet) and then linking this database with other databases (notes), and create views. This is actually quite similar to the board view, I guess. Every entry is a note (like in notion)

For trilium private use cases, these linked databases are not toooo useful and for special use cases one might create some kind of js-backend or something.

That said, being able to ditch google sheets for trilium could be nice.

**Comment by kalmigs** [comment]
@thfrei What I want is basically Excel or google sheets inside trilium, maybe a way to link or extract or show cell values outside of the spreadsheet to the trilium notes,

I saw in a different discussion that requested an `SQLite note`, might be overkill but yeah what you described is something cool to have too

**Comment by eeshaa1** [comment]
The absence of proper spreadsheets in Trillium is glaring. Trillium aims to be a lot more "data-centric" than Obsidian, so-to-speak, so it is strange that Obsidian has real spreadsheets while Trillium does not

**Comment by eliandoran** [comment]
@eeshaa1 , can you provide more info about Obsidian's real spreadsheet?

**Comment by eeshaa1** [comment]
@eliandoran Obsidian has many plugins that let you implement more feature-rich spreadsheets in various ways ((one method is literally just importing a google sheets "instance" (is that the term for it?) as a note))

**Comment by eliandoran** [comment]
@eeshaa1 , can you provide a link to those plugins so that I can analyze them?

Google sheets is not really embeddedable, as far as I know. If it's just an `<iframe>`/embedded browser, then you can achieve the same in Trilium using [Web Views](https://docs.triliumnotes.org/user-guide/note-types/webview).

Either way, plugins are third-party content, so saying that Obsidian has real spreadsheets while Trilium does not is not technically correct. Theoretically, someone can implement spreadsheets in Trilium as well (although clearly a more defined plugin system would encourage more developers). When something is part of the core, we are talking about more than just a spreadsheet, we are talking about data security and long-term maintenance. Plugins don't go through the same amount of scrutiny.

**Comment by thfrei** [comment]
I actually did research a little bit. 
The most popular obisidian spreadsheet plugin seem to be:
- [Obsidian Sheet Plus](https://github.com/ljcoder2015/obsidian-sheet-plus) which uses univer, a mixed open source free and licensed software. Quite capable, I think.
- [Obsidian Spreadsheets Plugin](https://github.com/divamgupta/obsidian-spreadsheets) which uses [Luckysheet ](https://github.com/dream-num/Luckysheet?tab=readme-ov-file) (which is NO longer maintained and instead recommends to use [univer](https://github.com/dream-num/univer)

I also think, that univer is the most suitable spreadsheet solution. However there are a few others. It seems, that the spreadsheet content could just be saved as a json.
One of the most useful use cases would then be to reference one cell somewhere in a trilium note. like {sheet-id}{A1} but of course that's a lot to integrate.

Minimal example with univer:
https://docs.univer.ai/showcase/sheets/slim-via-preset

```
import { UniverSheetsCorePreset } from '@univerjs/preset-sheets-core'
import sheetsCoreEnUS from '@univerjs/preset-sheets-core/locales/en-US'
import { createUniver, LocaleType, mergeLocales } from '@univerjs/presets'
import { WORKBOOK_DATA } from './data'

import './styles.css'

import '@univerjs/preset-sheets-core/lib/index.css'

const { univerAPI } = createUniver({
  locale: LocaleType.EN_US,
  locales: {
    [LocaleType.EN_US]: mergeLocales(
      sheetsCoreEnUS,
    ),
  },
  presets: [
    UniverSheetsCorePreset({
      container: 'app',
    }),
  ],
})

univerAPI.createWorkbook(WORKBOOK_DATA)
```

**Comment by eliandoran** [comment]
@thfrei , thanks for the analysis.

I looked into Univer as well and I'm aware it's probably one of the only solutions that is actually functional. The only problem is that I can't seem to find information about who is behind the development of that project and how safe would it be to use long-term.

For example, we are using for the Table collection a library called Tabulator which I think is really great. But turns out it has entered maintenance mode (see https://github.com/olifolkerd/tabulator/issues/4823#issuecomment-3740171567).

If we have to switch to another table solution for the collections, it's a major annoyance but not critical. But imagine the impact it would have for us to have to switch over to something else, when user data is already involved.

That's why probably the spreadsheet functionality will only be implemented as a plugin, with assurance of bitrot protection.

**Comment by eliandoran** [comment]
@thfrei ,

I have decided to implement Univer Sheets as a new note type for now, with clear disclaimer about the limitations.
After we implement a plugin system, we might extract it into a core plugin.
First version will be available in the next release, [v0](https://github.com/TriliumNext/Trilium/pull/8902) will be available in tomorrow's nightly, if you guys are interested. :)

Feel free to offer feedback or request some particular plugins to integrate (for now it's barebones).

Note that the sync version will be increased to avoid server version mismatches which will turn the spreadsheets into files.

**Comment by thfrei** [comment]
I like it! And it works :-)
I was actually thinking, that it can wait, but since you implemented it, I thought I give it a try. So here is my first dump. Sorry if it's messy.

From testing i found no bugs yet, just features (that you probably know)
- [ ] read-only mode is missing (if desired https://docs.univer.ai/showcase/sheets/read-only)
  - [ ] note revisions could use read-only mode
  - [ ] Same for shared notes read-only
- 🐞? maybe univer issue: when i type in a formula, like a sum for a few fields, the active cell jumps to A1 on the first sheet (not always?) - pattern when maybe more than 5 fields or so in sum)

https://github.com/user-attachments/assets/cc6c9105-e37a-4ccf-a471-f882fa69f44a

- [ ] I guess we need to remove the protect worksheet and person functionality?

<img width="1259" height="1019" alt="Image" src="https://github.com/user-attachments/assets/4a1c7015-d3ef-4f4e-aefe-ae04e4bb8f6b" />

- ✅ trilium tabs content sync work
- [ ] Column width 0 `The column width is less than 0, need to adjust page width to make it great than 0`
  - I googled a bit, with AI suggest requestAnimationFrame something (it removes at least the error message, I guess it's a rendering thing with react and univer, I'm not that into the rendering that I could say if this is a nice solution or not)
```
function useInitializeSpreadsheet(containerRef: MutableRef<HTMLDivElement | null>, apiRef: MutableRef<FUniver | undefined>) {
    useEffect(() => {
        if (!containerRef.current) return;

        let frameId: number;
        let isMounted = true;

        // Use requestAnimationFrame to ensure the container is painted before initializing Univer
        frameId = requestAnimationFrame(() => {
            if (!isMounted || !containerRef.current) return;

            const { univerAPI } = createUniver({
                locale: LocaleType.EN_US,
                locales: {
                    [LocaleType.EN_US]: mergeLocales(
                        UniverPresetSheetsCoreEnUS
                    ),
                },
                presets: [
                    UniverSheetsCorePreset({
                        container: containerRef.current,
                    })
                ]
            });

            apiRef.current = univerAPI;
        });

        return () => {
            isMounted = false;
            cancelAnimationFrame(frameId);

            if (apiRef.current) {
                apiRef.current.dispose();
                apiRef.current = undefined;
            }
        };

    }, [ apiRef, containerRef ]);
}
```

Ideas:
- I'd really like charts..., because that's what i use spreadsheets for, to generate charts. https://docs.univer.ai/showcase/sheets/charts
- Request single cell values, or charts from other text notes (or cross workbook? https://docs.univer.ai/showcase/sheets/cross-workbook-formula)
- search and replace inside note (however same shortcut?) https://docs.univer.ai/showcase/sheets/find-replace
- I also love annotations: https://docs.univer.ai/showcase/sheets/notes

Well it gets big fast... 
Thank you for your continous work!

Best wishes
Thomas

**Comment by eliandoran** [comment]
@thfrei , thanks for your feedback, I'll look into it.

> I'd really like charts..., because that's what i use spreadsheets for, to generate charts. https://docs.univer.ai/showcase/sheets/charts

Unfortunately this one won't work because it's a pro plugin. :(
https://docs.univer.ai/guides/sheets/features/charts

>  Same for shared notes read-only

Actually there is a small issue with sharing. More specifically, it is a standalone mode so we can't simply embed Univer in it. Although theoretically we can, we'd duplicate the libraries used since the share functionality is a separate application.

Another problem is the lack of export features (for example to HTML for printing).

**Comment by thfrei** [comment]
Ah yes..., this is unfortunate. However, checking https://univer.ai/pro it seems, that one could use basic charts even though then a watermark or something would be present. Import/Export is possible as well, albeit with limits. And 1 pivot table per document 😅
It is not super nice, but something (I guess)

**Comment by DigitallyGray** [comment]
@eliandoran , hello, I tested out the spreadsheets, and they work very well. I personally wouldn't need them for anything complex. So, beyond the already discussed charts, I don't have any big feature requests. But I would like to raise a small UX improvement/ consistency, if it hasn't already been noticed.

I did a few small recordings to show what i mean:
https://github.com/user-attachments/assets/e3888454-ce79-4856-b3d3-d707ee163d71
vs
https://github.com/user-attachments/assets/f005bb26-0573-4754-a0c6-dd290b435b3d
Also, just to clarify, there are no issues with Drag selection and Shift+Arrow. Only occurred to me now that including them implied an issue, just wanted to show the full range of behavior, just in case.

(Apologies in advance if there are issues with the attachments, I am not overly used to GitHub. If needed I can describe the issue.)

Thank you for your work on TriliumNotes.

**Comment by eliandoran** [comment]
Hi, @DigitallyGray .

Thanks for raising this issue.

I've looked a bit into it and the mismatch is because we are using a third-party library to handle the spreadsheets (Univer). The spreadsheet, for performance reasons is using canvas to draw the spreadsheet, including the editing experience as far as I know. As such the common cursor navigation won't work (from what I see not even Home or End). 

Should you wish to get this fixed, most likely this will have to be reported [upstream](https://github.com/dream-num/univer/issues). I see there were some fixes in the past regarding Shift to select (see https://github.com/dream-num/univer/issues/354 and https://github.com/dream-num/univer/pull/1217).

**Comment by DigitallyGray** [comment]
Thank you @eliandoran  for the quick response. I'll look to see if there has been an issue raised for it over on Univer regarding Ctrl+Arrow and Double Click Drag later, and if not raise one.

There is a small bug I encountered and thought I'd raise it here just in case it could be fixed here downstream. I've attached a screen record of it.

https://github.com/user-attachments/assets/84aa52ad-c7bb-4ac0-aaeb-0dc90f03dc86

**Comment by eliandoran** [comment]
@DigitallyGray , that does indeed seem to be an issue on our side, I'll investigate.

@thfrei ,
In the meantime, the latest nightly should contain more features: annotations, data validation, filtering, sorting, find/replace.
For now, no premium features such as charts, we might implement them in a later version.
